pkg/controller/dns/resourcerecordset.go

```diff
@@ -0,0 +1,148 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-gcp/apis/database/v1beta1"
+	"github.com/crossplane/provider-gcp/apis/dns/v1alpha1"
+	gcp "github.com/crossplane/provider-gcp/pkg/clients"
+	rrs "github.com/crossplane/provider-gcp/pkg/clients/resourcerecordset"
+)
+
+// Error strings.
+const (
+	errNewClient    = "cannot create new GCP DNS API client"
+	errNotRecordSet = "managed resource is not a DNS ResourceRecordSet"
+	errGet          = "cannot get GCP ResourceRecordSet object via DNS API"
+	errCreate       = "cannot create GCP ResourceRecordSet object via DNS API"
+	errUpdate       = "cannot update GCP ResourceRecordSet object via DNS API"
+	errDelete       = "cannot delete GCP ResourceRecordSet object via DNS API"
+)
+
+// SetupResourceRecordSet adds a controller that reconciles
+// ResourceRecordSet managed resources.
+func SetupResourceRecordSet(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1beta1.CloudSQLInstanceGroupKind)
+
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1beta1.CloudSQLInstanceGroupVersionKind),
+		managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+		managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), managed.NewNameAsExternalName(mgr.GetClient())),
+		managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.ResourceRecordSet{}).
+		Complete(r)
+}
+
+type connecter struct {
+	client client.Client
+}
+
+// Connect sets up iam client using credentials from the provider
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	projectID, opts, err := gcp.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+
+	cl, err := rrs.NewClient(ctx, projectID, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &external{client: cl}, errors.Wrap(err, errNewClient)
+}
+
+type external struct {
+	client rrs.Client
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotRecordSet)
+	}
+
+	fromProvider, err := e.client.Get(ctx, cr)
+	if gcp.IsErrorNotFound(err) {
```

Comment:
Is it possible to use this pattern here instead? https://github.com/crossplane/provider-gcp/blob/master/pkg/controller/pubsub/topic.go#L103